src/module/actor/character/index.ts

```diff
@@ -2166,15 +2172,24 @@ class CharacterPF2e extends CreaturePF2e {
 
     protected override async _preUpdate(
         changed: DeepPartial<CharacterSource>,
-        options: DocumentModificationContext<this>,
+        options: CreatureUpdateContext<this>,
         user: UserPF2e
     ): Promise<void> {
         const systemData = this.data.data;
 
         // Clamp level, allowing for level-0 variant rule and enough room for homebrew "mythical" campaigns
         const level = changed.data?.details?.level;
-        if (level?.value !== undefined) {
+        if (typeof level?.value === "number") {
             level.value = Math.clamped(Number(level.value) || 0, 0, 30);
+
+            // Adjust hit points if level is changing
+            const clone = this.clone(changed);
+            const hpMaxDifference = clone.hitPoints.max - this.hitPoints.max;
+            if (hpMaxDifference !== 0) {
+                options.allowHPOverage = true;
+                const newHitPoints = this.hitPoints.value + hpMaxDifference;
```

Comment:
this should probably be clamped to 1 hp if current hp is positive, and stay at 0 hp if its currently 0 hp.